flow/FastAlloc.cpp

```diff
@@ -224,25 +224,35 @@ struct FastAllocator<Size>::GlobalData {
 	long long totalMemory;
 	long long partialMagazineUnallocatedMemory;
 	long long activeThreads;
-	GlobalData() : totalMemory(0), partialMagazineUnallocatedMemory(0), activeThreads(0) { 
+	GlobalData() : totalMemory(0), partialMagazineUnallocatedMemory(0), activeThreads(0) {
 		InitializeCriticalSection(&mutex);
 	}
 };
 
 template <int Size>
 long long FastAllocator<Size>::getTotalMemory() {
-	return globalData()->totalMemory;
+	EnterCriticalSection(&globalData()->mutex);
+	long long total = globalData()->totalMemory;
+	LeaveCriticalSection(&globalData()->mutex);
```

Comment:
For simple reads (not sure about the potentially-torn arithmetic involving `size()`), the TSAN errors are false positives on x86 and arm, at least.  In particular, at mutex release, the compiler spills registers for potentially shared state, and the processor flushes the write buffer (speaking in x86 terms), so the write is guaranteed to be globally visible once the mutex is released.  If the value is read while the mutex is held, readers will get the old value or the new value depending on whether the write was flushed for other reasons.  ARM's semantics are a bit weaker, but shouldn't cause any trouble here.

Grabbing the mutexes incurs a bit of a performance hit, but (worse) unnecessary synchronization logic is a maintenance nightmare (Why is this useless-looking mutex here?  Can I delete it?)  It's probably better to use `MOZ_TSAN_BLACKLIST`(and add a comment) to tell TSAN these functions are OK.  `std::atomic` would also be OK with me, since it's not as confusing (but it's still unnecessary).

It's known to be impossible to implement a sound and complete race detection algorithm, so it's not surprising TSAN is giving false positives.  (Race detection is an ill-posed problem in the first place; I could go on for hours...)

Reply:
> the TSAN errors are false positives on x86 and arm

A data race in c++ has [a precise definition](https://en.cppreference.com/w/cpp/language/memory_model#Threads_and_data_races), and this is a data race.

> ARM's semantics are a bit weaker, but shouldn't cause any trouble here.

We're not writing ARM code, so ARM's semantics aren't relevant here. c++ semantics are relevant.

> std::atomic would also be OK with me, since it's not as confusing

This would also be ok with me.

>  unnecessary synchronization logic is a maintenance nightmare (Why is this useless-looking mutex here? Can I delete it?) It's probably better to use MOZ_TSAN_BLACKLIST(and add a comment) to tell TSAN these functions are OK.

A TSAN suppression would be a maintenance nightmare, especially since this mutex is not useless.

> so it's not surprising TSAN is giving false positives.

TSAN shouldn't have false positives[1] (assuming all code was instrumented by tsan)


[1]:

> TSAN  V2  implements  a  pure  happens-before  algo-rithm, and hence has no false positives.

From https://static.googleusercontent.com/media/research.google.com/en//pubs/archive/43217.pdf

Reply:
Section 4.3 of the TSAN V2 paper says that some synchronization patterns are inexpressible in their ThreadSafety model, and that others need to be rewritten "orthogonally" to readability.  From the perspective of thread safety, those are false positives.

Anyway, C++ defines a cleverly chosen class of data races as undefined behavior, and if they guarantee no false positives vs. that definition, that's useful.

Do you have a preference for the load / store ordering semantics?  `memory_order_relaxed` should suffice for this use case.  It guarantees atomicity, but not ordering.

The stronger orderings are defined in terms of data dependencies between threads, but there aren't any data dependencies with other variables in the functions in question.  `relaxed` isn't really doing anything different than the current code (it doesn't emit any thread synchronization instructions on architectures we care about).  The main difference is  that we're bound to eventually hit a data dependency in some unrelated code, and that'll force the optimizer to load/store from the heap.  In practice, the optimizer will do that anyway already because it'll eventually be forced to spill the register.

`memory_order_relaxed` is closest to the intent of this code, so I'd find it the most readable, but I don't have a strong opinion.

Reply:
`memory_order_relaxed` sgtm from my reading of https://en.cppreference.com/w/cpp/atomic/memory_order#Relaxed_ordering

> `memory_order_relaxed` is closest to the intent of this code, so I'd find it the most readable

I think that makes sense. I'd usually just go with the default `memory_order_seq_cst` since I'm most familiar with that. I think which one I find more readable depends on how recently I've read https://en.cppreference.com/w/cpp/atomic/memory_order, so I also don't have a strong opinion.

Reply:
I tend to leave off the parameter (which defaults to memory_order_seq_cst) unless I've thought carefully about it.  (Again, for documentation purposes more than the performance win.)

Thanks.